enhancement: allow ocis to provide custom web applications (#8523)
owncloud/ocis: services/web/README.md

### Embedding Web

Web can be consumed by another application in a stripped down version called “Embed mode”. This mode is supposed to be used in the context of selecting or sharing resources. For more details see the developer documentation [ownCloud Web / Embed Mode](https://owncloud.dev/clients/web/embed-mode/). See the environment variables: `WEB_OPTION_MODE` and `WEB_OPTION_EMBED_TARGET` to configure the embedded mode.

# Web Apps

The administrator of the environment is capable of providing custom web applications to the users.
This feature is useful for organizations that want to provide third party or custom apps to their users.

It's important to note that the feature at the moment is only capable of providing static (js, mjs, e.g.) web applications
and does not support injection of dynamic web applications (custom dynamic backends).

## Loading Applications

Loading applications is achieved in three ways:

* By loading a web application from a user-provided path, by setting the `WEB_ASSET_APPS_PATH` environment variable.

> * By loading a web application from the default ocis home directory, e.g. `$OCIS_BASE_DATA_PATH/web/assets/apps`.

Reviewer: While the first sentence is clear as you define a path in a variable, the second sentence is not clear.
The value provided is a path but where to configure it?

Author: $OCIS_BASE_DATA_PATH is our default ocis home path, i have no idea how to describe it better, can you provide a better explaining version?

Reviewer: We have solved this via chat and you commit added

* By embedding a web application into the ocis binary which is a build-time option,
  e.g. `ocis_src_path/services/web/assets/apps` followed by a build.

The list of available applications is composed of the build in extensions and the custom applications
provided by the administrator, e.g. `WEB_ASSET_APPS_PATH` or `$OCIS_BASE_DATA_PATH/web/assets/apps`.

> for example, if ocis contains a build in extension `image-viewer-dfx` and the administrator provides a custom
> application `image-viewer-obj` in the `WEB_ASSET_APPS_PATH` directory,the user will be able to access both applications
> from the web ui.

Reviewer: ```suggestion
For example, if ocis would contain a build in extension named `image-viewer-dfx` and the administrator provides a custom application named `image-viewer-obj` in the `WEB_ASSET_APPS_PATH` directory, the user will be able to access both applications from the web ui.
```

## Application Structure

Applications always have to follow a strict structure, which is as follows:

> * an application must be a directory
> * an application must contain a `manifest.json` file

Reviewer: ```suggestion
* each application must be in its own directory
* each application directory must contain a `manifest.json` file
```

> everything else is skipped and not considered as an application.

Reviewer: ```suggestion
Everything else is skipped and not considered as an application.
```

```json
{
  "entrypoint": "index.js",
  "config": {
    "maxWith": 1280,
    "maxHeight": 1280
  }
}
```

and the `apps.yaml` file contains the following configuration:

```yaml
image-viewer-obj:
  config:
    maxHeight: 640
    maxSize: 512
```

the final configuration for web will be:

```json
{
  "external_apps": [
    {
      "id": "image-viewer-obj",
      "path": "index.js",
      "config": {
        "maxWith": 1280,
        "maxHeight": 640,
        "maxSize": 512
      }
    }
  ]
}
```

besides the configuration from the `manifest.json` file, the `apps.yaml` file can also contain the following fields:

* `disabled` - optional - defaults to `false` - if set to `true`, the application will not be loaded

The local provided configuration yaml will always override the shipped application manifest configuration.

## Fallback Mechanism

Besides the configuration and application registration, there is one further important aspect to know;
in the process of loading the application assets, the system uses a fallback mechanism to load the assets.

This is incredibly useful for cases where just a single asset should be overwritten, e.g., a logo or similar.

> Consider the following, ocis is shipped with a default extension called `image-viewer-dfx` which contains a logo,

Reviewer: ```suggestion
Consider the following, ocis is shipped with a default extension named `image-viewer-dfx` which contains a logo,
```

### Embedding Web

Web can be consumed by another application in a stripped down version called “Embed mode”. This mode is supposed to be used in the context of selecting or sharing resources. For more details see the developer documentation [ownCloud Web / Embed Mode](https://owncloud.dev/clients/web/embed-mode/). See the environment variables: `WEB_OPTION_MODE` and `WEB_OPTION_EMBED_TARGET` to configure the embedded mode.

# Web Apps

The administrator of the environment is capable of providing custom web applications to the users.

> This feature is useful for organizations that have specific web applications that they want to provide to their users.

Reviewer 2: ```suggestion
This feature is useful for organizations that want to provide third party or custom apps to their users.
```

It's important to note that the feature at the moment is only capable of providing static (js, mjs, e.g.) web applications
and does not support injection of dynamic web applications (custom dynamic backends).

## Loading Applications

> Loading applications is provided in three ways:

Reviewer 2: ```suggestion
Loading applications is achieved in three ways:
```

* By loading a web application from a user-provided path, by setting the `WEB_APPS_PATH` environment variable.
* By loading a web application from the default ocis home directory, e.g. `$OCIS_BASE_DATA_PATH/web/assets/apps`.
* By embedding a web application into the ocis binary which is a build-time option,
  e.g. `ocis_src_path/services/web/assets/apps` followed by a build.

The list of available applications is composed of the build in extensions and the custom applications
provided by the administrator, e.g. `WEB_APPS_PATH` or `$OCIS_BASE_DATA_PATH/web/assets/apps`.

for example, if ocis contains a build in extension `image-viewer-dfx` and the administrator provides a custom
application `image-viewer-obj` in the `WEB_APPS_PATH` directory,the user will be able to access both applications
from the web ui.

## Application Structure

Applications always have to follow a strict structure, which is as follows:

* an application must be a directory
* an application must contain a `manifest.json` file

everything else is skipped and not considered as an application.

> The `manifest.json` file contain the following fields:

Reviewer 3: ```suggestion
The `manifest.json` file contains the following fields:
```

> * `id` - required - the name of the application must be unique across all applications

Reviewer 2: Could you please add where the id is coming from? Does it need to be equal to the folder name in the apps asset path? Or same id as in the appInfo object inside the entrypoint?

Edit: found the code path. Id needs to be equal to the folder name. Wouldn't it make more sense to read the folder name from the filesystem then and don't include the field in the manifest.json at all?

* `entrypoint` - required - the entrypoint of the application, e.g. `index.js`, the path is relative to the parent directory
* `config` - optional - a list of key-value pairs that are passed to the global web application configuration

## Application Configuration

> it's important to note that an application manifest should never be changed manually;

Reviewer 3: ```suggestion
It's important to note that an application manifest should never be changed manually;
```

`$OCIS_BASE_DATA_PATH/config/apps.yaml` file.

The `apps.yaml` file must contain a list of key-value pairs which gets merged with the `config` field.
For example, if the `image-viewer-obj` application contains the following configuration:

```json
{
  "id": "image-viewer-obj",
  "entrypoint": "index.js",
  "config": {
    "maxWith": 1280,
    "maxHeight": 1280
  }
}
```

and the `apps.yaml` file contains the following configuration:

```yaml
image-viewer-obj:
  config:
    maxHeight: 640
    maxSize: 512
```

the final configuration for web will be:

```json
{
  "external_apps": [
    {
      "id": "image-viewer-obj",
      "path": "index.js",
      "config": {
        "maxWith": 1280,
        "maxHeight": 640,
        "maxSize": 512
      }
    }
  ]
}
```

besides the configuration from the `manifest.json` file, the `apps.yaml` file can also contain the following fields:

* `disabled` - optional - defaults to `false` - if set to `true`, the application will not be loaded

The local provided configuration yaml will always override the shipped application manifest configuration.

## Fallback Mechanism

> Besides the configuration and application registration, there is one further important aspect to now;

Reviewer 3: ```suggestion
Besides the configuration and application registration, there is one further important aspect to know;
```
